src/client/pool/singleton.rs

```diff
@@ -0,0 +1,333 @@
+//! Singleton pools
+//!
+//! The singleton pool combines a MakeService that should only produce a single
+//! active connection. It can bundle all concurrent calls to it, so that only
+//! one connection is made. All calls to the singleton will return a clone of
+//! the inner service once established.
+//!
+//! This fits the HTTP/2 case well.
+
+use std::sync::{Arc, Mutex};
+use std::task::{self, Poll};
+
+use tokio::sync::oneshot;
+use tower_service::Service;
+
+use self::internal::{DitchGuard, SingletonError, SingletonFuture};
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// A singleton pool over an inner service.
+///
+/// The singleton wraps an inner service maker, bundling all calls to ensure
+/// only one service is created. Once made, it returns clones of the made
+/// service.
```

Comment:
i'd suggest that we clarify this slightly, and indicate that we _lazily_ instantiate the inner service.

i could imagine an alternative "eager" singleton connection pool that invokes the `MakeService` immediately, for example.